chore(blockifier): create unit tests for contract_class_manager

avivg-starkware opened the pull request.

reviewable-StarkWare commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/starkware-libs/sequencer/2768)
<!-- Reviewable:end -->

avivg-starkware commented:
* **#2768** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2768?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2768?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

noaov1 reviewed (changes requested):
> Reviewed 7 of 7 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 7 unresolved discussions (waiting on @avi-starkware and @avivg-starkware)

___
*[`crates/blockifier/src/test_utils/contracts.rs` line 183 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSHVtx7elcTWPxnsLf:-OFSHVtx7elcTWPxnsLg:b-4kpyeo) ([raw file](https://github.com/starkware-libs/sequencer/blob/445cd55ed11934615e77fe905da6377ada7dd601/crates/blockifier/src/test_utils/contracts.rs#L183)):*
> ```XML
>     }
> 
>     pub fn get_casm(&self) -> CompiledClassV1 {
> ```

Consider sharing code with`get_class`.



_Code quote:_
```XML
pub fn get_casm(&self) -> CompiledClassV1 {
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 19 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSI-bE09exnTqppIDh:-OFSI-bE09exnTqppIDi:b-8k23n5) ([raw file](https://github.com/starkware-libs/sequencer/blob/445cd55ed11934615e77fe905da6377ada7dd601/crates/blockifier/src/state/contract_class_manager_test.rs#L19)):*
> ```XML
>     let mut contract_class = test_contract.get_contract_class();
>     // Truncate the sierra program to trigger an error.
>     contract_class.sierra_program = contract_class.sierra_program[..100].to_vec();
> ```

Nice :)
Is there a way to cause a compilation error by "damaging" the `SierraContractClass` object?



_Code quote:_
```XML
    // Truncate the sierra program to trigger an error.
    contract_class.sierra_program = contract_class.sierra_program[..100].to_vec();
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 77 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSEvwbBrLMdaUnvmSQ:-OFSEvwbBrLMdaUnvmSR:bimnmyo) ([raw file](https://github.com/starkware-libs/sequencer/blob/445cd55ed11934615e77fe905da6377ada7dd601/crates/blockifier/src/state/contract_class_manager_test.rs#L77)):*
> ```XML
>         "Sender should be able to send a request successfully"
>     );
> }
> ```

Consider unifying these tests with `test_cases`. 
You can also check: no compiler when cairo-native is off. 



_Code quote:_
```XML
#[test]
fn test_sender_with_native_compilation_disabled() {
    let config = ContractClassManagerConfig { run_cairo_native: false, ..Default::default() };
    let manager = ContractClassManager::start(config);
    assert!(manager.sender.is_none(), "Sender should be None when native compilation is disabled");
}

#[test]
fn test_sender_with_native_compilation_enabled() {
    let config = ContractClassManagerConfig { run_cairo_native: true, ..Default::default() };
    let manager = ContractClassManager::start(config);
    assert!(manager.sender.is_some());

    assert!(
        manager.sender.as_ref().unwrap().try_send(create_test_request()).is_ok(),
        "Sender should be able to send a request successfully"
    );
}
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 80 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSFvXp3C5qbhK3dWHU:-OFSFvXp3C5qbhK3dWHV:b-icozcc) ([raw file](https://github.com/starkware-libs/sequencer/blob/445cd55ed11934615e77fe905da6377ada7dd601/crates/blockifier/src/state/contract_class_manager_test.rs#L80)):*
> ```XML
> 
> #[test]
> fn test_send_request_channel_disconnected() {
> ```

Can you please explain this test?



_Code quote:_
```XML
fn test_send_request_channel_disconnected() 
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 108 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSGCEF7hmRbWO_QA9m:-OFSGCEF7hmRbWO_QA9n:b-k4jhub) ([raw file](https://github.com/starkware-libs/sequencer/blob/445cd55ed11934615e77fe905da6377ada7dd601/crates/blockifier/src/state/contract_class_manager_test.rs#L108)):*
> ```XML
>     let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
>         manager.send_compilation_request(second_request);
>     }));
> ```

Why is the `catch_unwind` needed?



_Code quote:_
```XML
    let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
        manager.send_compilation_request(second_request);
    }));
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 114 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSGWvR9fGFhM4TzBpo:-OFSGWvR9fGFhM4TzBpp:b3kre) ([raw file](https://github.com/starkware-libs/sequencer/blob/445cd55ed11934615e77fe905da6377ada7dd601/crates/blockifier/src/state/contract_class_manager_test.rs#L114)):*
> ```XML
> 
> #[test]
> fn test_run_compilation_worker_success() {
> ```

?



_Suggestion:_
```XML
fn test_process_compilation_request_success() {
```
___
*[`crates/blockifier/src/state/contract_class_manager.rs` line 33 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSDjwcA04z_7zwD_sW:-OFSDjwd1k8ccYhTl6ob:b-3a0hu7) ([raw file](https://github.com/starkware-libs/sequencer/blob/445cd55ed11934615e77fe905da6377ada7dd601/crates/blockifier/src/state/contract_class_manager.rs#L33)):*
> ```XML
> pub const DEFAULT_COMPILATION_REQUEST_CHANNEL_SIZE: usize = 1000;
> 
> #[cfg(all(test, feature = "cairo_native"))]
> ```

We will want to test it even when the `cairo-native` flag is off.



_Code quote:_
```XML
#[cfg(all(test, feature = "cairo_native"))]
```


<!-- Sent from Reviewable.io -->

avivg-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 19 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSI-bE09exnTqppIDh:-OFW5RwL2A0vGvemcE_L:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/contract_class_manager_test.rs#L19)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Nice :)
Is there a way to cause a compilation error by "damaging" the `SierraContractClass` object?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

avivg-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 77 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSEvwbBrLMdaUnvmSQ:-OFW8xmy1V6Jzua5bXqS:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/contract_class_manager_test.rs#L77)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Consider unifying these tests with `test_cases`. 
You can also check: no compiler when cairo-native is off.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

avivg-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 80 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSFvXp3C5qbhK3dWHU:-OFW9_Dc2Ulx4kiDDGAZ:b-sb1g6t) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/contract_class_manager_test.rs#L80)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Can you please explain this test?
</blockquote></details>

my intention was to check that panic occurs when trying to send massage with disconnected channel, how does it look now? 


<!-- Sent from Reviewable.io -->

avivg-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 108 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSGCEF7hmRbWO_QA9m:-OFaaAun3FEVtw5KB3Vg:b-nnm3nn) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/contract_class_manager_test.rs#L108)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Why is the `catch_unwind` needed?
</blockquote></details>

WDYT? not sure if this test is needed and if so, what best way to handle a check that panic doesn't occur


<!-- Sent from Reviewable.io -->

avivg-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 114 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSGWvR9fGFhM4TzBpo:-OFaaWQf-NUXEULnVk_o:b-glqirz) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/contract_class_manager_test.rs#L114)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

?
</blockquote></details>

I've changed it to test send_compilation_request and added a separate test for process_compilation_request.
Do you think we need a test for the case wait_on_native = false? using time funcs to ensure enough time elapses for the other thread to finish compiling 


<!-- Sent from Reviewable.io -->

meship-starkware reviewed (commented):
> Reviewed 2 of 7 files at r2, 1 of 1 files at r3, 1 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 6 of 7 files reviewed, 8 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @noaov1)

___
*[`crates/blockifier/src/state/global_cache.rs` line 33 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFVksl15p4qLxdXcDrp:-OFVksl15p4qLxdXcDrq:bdkeyzo) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/global_cache.rs#L33)):*
> ```XML
> 
> #[cfg(feature = "cairo_native")]
> #[derive(Debug, Clone, PartialEq)]
> ```

Can you derive the PartialEq only for the test config?



_Code quote:_
```XML
#[derive(Debug, Clone, PartialEq)]
```


<!-- Sent from Reviewable.io -->

avivg-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 5 of 7 files reviewed, 7 unresolved discussions (waiting on @avi-starkware, @meship-starkware, and @noaov1)

___
*[`crates/blockifier/src/state/global_cache.rs` line 33 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFVksl15p4qLxdXcDrp:-OFrMELy-OVBk_kGvKN4:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/global_cache.rs#L33)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

Can you derive the PartialEq only for the test config?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

avivg-starkware commented:
*[`crates/blockifier/src/test_utils/contracts.rs` line 183 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSHVtx7elcTWPxnsLf:-OFw7ltV2Rjpu0LFUHyo:bswanhp) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/test_utils/contracts.rs#L183)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Consider sharing code with`get_class`.
</blockquote></details>

handled code sharing in  #3138 


<!-- Sent from Reviewable.io -->

noaov1 reviewed (commented):
> Reviewed 1 of 11 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 3 of 15 files reviewed, 5 unresolved discussions (waiting on @avi-starkware and @meship-starkware)



<!-- Sent from Reviewable.io -->

meship-starkware reviewed (commented):
> Reviewed 2 of 7 files at r2, 8 of 11 files at r17, 1 of 1 files at r20.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 12 of 16 files reviewed, 6 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @noaov1)

___
*[`crates/blockifier/src/test_utils.rs` line 175 at r20](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OGV285yDmilNj3IFvkg:-OGV285z1qDKimJ5eAJg:b683wjc) ([raw file](https://github.com/starkware-libs/sequencer/blob/8031a24aaa5637db88f03270dc7f0dbb83c9db16/crates/blockifier/src/test_utils.rs#L175)):*
> ```XML
> }
> 
> impl TestLogger {
> ```

Not sure it should be in test utils, I can add a logger.rs file to the test_utils folder if you think its better 



_Code quote:_
```XML
pub struct TestLogger {
    logs: Arc<Mutex<Vec<String>>>,
}

impl Default for TestLogger {
    fn default() -> Self {
        Self::new()
    }
}

impl TestLogger {
```


<!-- Sent from Reviewable.io -->

meship-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 11 of 16 files reviewed, 6 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @noaov1)

___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 114 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSGWvR9fGFhM4TzBpo:-OGZs73B6jHbTk5HrcCY:b-4vw2uy) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/contract_class_manager_test.rs#L114)):*
<details><summary><i>Previously, avivg-starkware wrote…</i></summary><blockquote>

I've changed it to test send_compilation_request and added a separate test for process_compilation_request.
Do you think we need a test for the case wait_on_native = false? using time funcs to ensure enough time elapses for the other thread to finish compiling
</blockquote></details>

For the process, there is no need to make a test without waiting. The flow is the same 
as for the send compilation request, I added a test without waiting on the native request that checks the logs to see that a compilation request was sent 


<!-- Sent from Reviewable.io -->

avi-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 187 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH28YJFDiQHAwnYKnQU:-OH28YJFDiQHAwnYKnQV:bvewmg) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L187)):*
> ```XML
>     let (request, native) = create_test_request_with_native();
>     sender.try_send(request.clone()).unwrap();
>     drop(sender);
> ```

Do you drop the sender so that the compilation will terminate after pulling one request from the channel?

Please document this line



_Code quote:_
```XML
    drop(sender);
```


<!-- Sent from Reviewable.io -->

avi-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 197 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH29CfYC-NdsBRYYeq7:-OH29CfYC-NdsBRYYeq8:btj4p84) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L197)):*
> ```XML
>     run_compilation_worker(contract_caches.clone(), receiver, compiler);
> 
>     let cached_native = manager.get_native(&request.0);
> ```

How do you know the compilation is done at this point?
Is `run_compilation_worker` blocking?



_Code quote:_
```XML
    let cached_native = manager.get_native(&request.0);
```


<!-- Sent from Reviewable.io -->

avi-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 88 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH3OywB7kcmZSsPTY_L:-OH3OywB7kcmZSsPTY_M:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L88)):*
> ```XML
>     } else {
>         assert!(manager.sender.is_none(), "Sender should be None");
>         assert!(manager.compiler.is_none(), "Compiler should be None");
> ```





_Suggestion:_
```XML
            assert!(manager.sender.is_none(), "Sender should be None - the compilation worker is not used.");
            assert!(manager.compiler.is_some(), "Compiler should be Some - compilation is not offloaded to the compilation worker.");
        } else {
            assert!(manager.sender.is_some(), "Sender should be Some - the compilation worker is used.");
            assert!(manager.compiler.is_none(), "Compiler should be None - compilation is offloaded to the compilation worker.");
        }
    } else {
        assert!(manager.sender.is_none(), "Sender should be None - Cairo native is disabled.");
        assert!(manager.compiler.is_none(), "Compiler should be None - Cairo native is disabled.");
```


<!-- Sent from Reviewable.io -->

avi-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 100 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH3Pc3lCYFNQmg9p4Kh:-OH3Pc3m-rPqbmIk4nRN:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L100)):*
> ```XML
>     let contract_caches = ContractCaches::new(config.contract_cache_size);
>     let (sender, receiver) = sync_channel(native_config.channel_size);
>     drop(receiver);
> ```





_Suggestion:_
```XML
    // Disconnect the channel by dropping the receiver.
    drop(receiver);
```


<!-- Sent from Reviewable.io -->

avi-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 138 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH3QDrDA1b22w0sjNRp:-OH3QDrE7iatTBVIJOZ2:bvvun57) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L138)):*
> ```XML
>     let native_config = CairoNativeRunConfig {
>         run_cairo_native: false,
>         wait_on_native_compilation: true,
> ```

We should check both `true` and `false` here.



_Code quote:_
```XML
        wait_on_native_compilation: true,
```


<!-- Sent from Reviewable.io -->

avi-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 123 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH3Qeax6rFl2s4rE_nl:-OH3Qeax6rFl2s4rE_nm:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L123)):*
> ```XML
>         ContractClassManagerConfig { cairo_native_run_config: native_config, ..Default::default() };
>     let manager = ContractClassManager::start(config);
>     let (request, native) = create_test_request_with_native();
> ```





_Suggestion:_
```XML
    let (request, expected_native) = create_test_request_with_native();
```


<!-- Sent from Reviewable.io -->

avi-starkware reviewed (commented):
> Reviewed 1 of 3 files at r16, 1 of 11 files at r17, 1 of 4 files at r24, 1 of 2 files at r25, 1 of 1 files at r26, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 15 of 16 files reviewed, 11 unresolved discussions (waiting on @avivg-starkware, @meship-starkware, and @noaov1)



<!-- Sent from Reviewable.io -->

avivg-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 114 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSGWvR9fGFhM4TzBpo:-OH71IM2AuNEBYb3jaQ8:b9hatl9) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/contract_class_manager_test.rs#L114)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

For the process, there is no need to make a test without waiting. The flow is the same 
as for the send compilation request, I added a test without waiting on the native request that checks the logs to see that a compilation request was sent
</blockquote></details>

@meship-starkware  could you point me to that test? I can't seem to find it (I'm looking at all changes r8- r26)


<!-- Sent from Reviewable.io -->

avivg-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 15 of 16 files reviewed, 12 unresolved discussions (waiting on @avi-starkware, @meship-starkware, and @noaov1)

___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 114 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSGWvR9fGFhM4TzBpo:-OH7LFW93P7dqvYCS05X:bpqxn2v) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/contract_class_manager_test.rs#L114)):*
<details><summary><i>Previously, avivg-starkware wrote…</i></summary><blockquote>

@meship-starkware  could you point me to that test? I can't seem to find it (I'm looking at all changes r8- r26)
</blockquote></details>

@meship-starkware  I see that in the following PR #3335  you rewrote the deleted func: test_send_compilation_request_channel_full()

for clarity, I'm returning the func here, could you delete it in the PR #3335 where it's rewritten ?
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 138 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH3QDrDA1b22w0sjNRp:-OH7NQt69IBP0DOx6VXt:b-3g3dmb) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L138)):*
<details><summary><i>Previously, avi-starkware (Avi Cohen) wrote…</i></summary><blockquote>

We should check both `true` and `false` here.
</blockquote></details>

see @meship-starkware 's following PR #3335 func test_send_compilation_request_positive_flow
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 187 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH28YJFDiQHAwnYKnQU:-OH7QuiW6mFRVeKN-53A:bh3h96s) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L187)):*
<details><summary><i>Previously, avi-starkware (Avi Cohen) wrote…</i></summary><blockquote>

Do you drop the sender so that the compilation will terminate after pulling one request from the channel?

Please document this line
</blockquote></details>

Yes. added doc
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 197 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH29CfYC-NdsBRYYeq7:-OH7R4M_3CMelIw57pUg:bs31ohd) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L197)):*
<details><summary><i>Previously, avi-starkware (Avi Cohen) wrote…</i></summary><blockquote>

How do you know the compilation is done at this point?
Is `run_compilation_worker` blocking?
</blockquote></details>

as I'm running run_compilation_worker here and not though the new thread, why might it not be done sequentially? 
perhaps the creation of manager here is redundant? and better to directly compare the caches?  



_Code snippet:_
```XML
let cached_native = contract_caches.get_native(&request.0);
    assert_eq!(
        cached_native.unwrap(),
        CachedCairoNative::Compiled(expected_native),
        "Cached Native class should match the expected."
    );
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 100 at r24](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH2ut74AuhlqNx6eeX6:-OH2ut75399UVACoU8CJ:b-bf4tc6) ([raw file](https://github.com/starkware-libs/sequencer/blob/568c4197031974b9cc85627f5938bce797d097d8/crates/blockifier/src/state/contract_class_manager_test.rs#L100)):*
> ```XML
>     let contract_caches = ContractCaches::new(config.contract_cache_size);
>     let (sender, receiver) = sync_channel(native_config.channel_size);
>     drop(receiver);
> ```

To add document
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 88 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH3OywB7kcmZSsPTY_L:-OH7KHIO7TRzRsdfJDyY:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L88)):*
> ```XML
>     } else {
>         assert!(manager.sender.is_none(), "Sender should be None");
>         assert!(manager.compiler.is_none(), "Compiler should be None");
> ```

Done.
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 100 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH3Pc3lCYFNQmg9p4Kh:-OH7T2RB4UNFdeoDMLbI:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L100)):*
> ```XML
>     let contract_caches = ContractCaches::new(config.contract_cache_size);
>     let (sender, receiver) = sync_channel(native_config.channel_size);
>     drop(receiver);
> ```

Done.
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 123 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH3Qeax6rFl2s4rE_nl:-OH7KYAa2FdlhQbxQoJL:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L123)):*
> ```XML
>         ContractClassManagerConfig { cairo_native_run_config: native_config, ..Default::default() };
>     let manager = ContractClassManager::start(config);
>     let (request, native) = create_test_request_with_native();
> ```

Done.


<!-- Sent from Reviewable.io -->

avi-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 217 at r28](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH8DA-TDJ6pk1pWHePY:-OH8DA-TDJ6pk1pWHePZ:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/831cd298dbf096b22774b4e868212c8b8fa1023a/crates/blockifier/src/state/contract_class_manager_test.rs#L217)):*
> ```XML
>     sender.try_send(request.clone()).unwrap();
>     // Close the channel to signal no more requests. The worker should terminate after processing
>     // all pending requests.
> ```





_Suggestion:_
```XML
    // Drop the sender to close the channel and signal no more requests. Since the channel is closed,
    // the compilation worker should terminate after processing all pending requests.
```


<!-- Sent from Reviewable.io -->

avi-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 197 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH29CfYC-NdsBRYYeq7:-OH8BYjaBw4ydthiaN4k:b-yfhzcj) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L197)):*
<details><summary><i>Previously, avivg-starkware wrote…</i></summary><blockquote>

as I'm running run_compilation_worker here and not though the new thread, why might it not be done sequentially? 
perhaps the creation of manager here is redundant? and better to directly compare the caches?
</blockquote></details>

No. I think it is okay.
Just explain in a comment that the function does not run in a separate thread, and is therefore blocking until it compiles all requests in the (disconnected) compilation request channel.


<!-- Sent from Reviewable.io -->

avi-starkware reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 14 files at r27, 1 of 1 files at r28, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-:-OH8LagQBaxsmg2x8YJ8:bnfp4nl)* status: 4 of 16 files reviewed, 9 unresolved discussions (waiting on @avivg-starkware, @meship-starkware, and @noaov1)



<!-- Sent from Reviewable.io -->

avivg-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 3 of 16 files reviewed, 8 unresolved discussions (waiting on @avi-starkware, @meship-starkware, and @noaov1)

___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 197 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH29CfYC-NdsBRYYeq7:-OH8RQvq3EzitZsXhMx5:bsm52ig) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L197)):*
<details><summary><i>Previously, avi-starkware (Avi Cohen) wrote…</i></summary><blockquote>

No. I think it is okay.
Just explain in a comment that the function does not run in a separate thread, and is therefore blocking until it compiles all requests in the (disconnected) compilation request channel.
</blockquote></details>

I was rethinking this, starting with changing the lines order as follows...
As the only use of manager is for accessing the caches I feel the creation of manager is mainly confusing rather than needed here. WDYT? sorry for the back and forth 



_Code snippet:_
```XML
    // Drop the sender to close the channel and signal no more requests.
    drop(sender);
    // Since the channel is closed, the compilation worker should terminate after processing all
    // pending requests.
    run_compilation_worker(contract_caches.clone(), receiver, compiler);
    
    let manager = ContractClassManager {
        cairo_native_run_config: native_config,
        contract_caches: contract_caches.clone(),
        sender: None,
        compiler: None,
    };

    assert_eq!(
        manager.get_native(&request.0).unwrap(),
        CachedCairoNative::Compiled(expected_native),
        "Cached Native class should match the expected."
    );
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 217 at r28](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH8DA-TDJ6pk1pWHePY:-OH8RJAX7SH3a0mFOc0L:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/831cd298dbf096b22774b4e868212c8b8fa1023a/crates/blockifier/src/state/contract_class_manager_test.rs#L217)):*
> ```XML
>     sender.try_send(request.clone()).unwrap();
>     // Close the channel to signal no more requests. The worker should terminate after processing
>     // all pending requests.
> ```

Done


<!-- Sent from Reviewable.io -->

avi-starkware commented:
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 197 at r26](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OH29CfYC-NdsBRYYeq7:-OH8wAq7938xzfXwwQUG:bsrfkwe) ([raw file](https://github.com/starkware-libs/sequencer/blob/dbdc2177da609689986a2a7ec9e4582a2fe7c61c/crates/blockifier/src/state/contract_class_manager_test.rs#L197)):*
<details><summary><i>Previously, avivg-starkware wrote…</i></summary><blockquote>

I was rethinking this, starting with changing the lines order as follows...
As the only use of manager is for accessing the caches I feel the creation of manager is mainly confusing rather than needed here. WDYT? sorry for the back and forth
</blockquote></details>

Looks good


<!-- Sent from Reviewable.io -->

meship-starkware reviewed (approved):
> Reviewed 1 of 3 files at r16, 1 of 11 files at r17, 1 of 14 files at r27, 13 of 13 files at r29, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 5 unresolved discussions (waiting on @avi-starkware and @noaov1)



<!-- Sent from Reviewable.io -->

meship-starkware reviewed (commented):
> Reviewed 4 of 4 files at r30.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed (commit messages unreviewed), 5 unresolved discussions (waiting on @avi-starkware and @noaov1)



<!-- Sent from Reviewable.io -->

meship-starkware reviewed (approved):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 5 unresolved discussions (waiting on @avi-starkware and @noaov1)



<!-- Sent from Reviewable.io -->

noaov1 reviewed (changes requested):
> Reviewed 1 of 3 files at r16, 1 of 4 files at r30, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 10 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @meship-starkware)

___
*[`crates/blockifier/src/state/global_cache.rs` line 33 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFVksl15p4qLxdXcDrp:-OHHiqGq47GTBIHpRQhj:b35x8zw) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/global_cache.rs#L33)):*
<details><summary><i>Previously, avivg-starkware wrote…</i></summary><blockquote>

Done
</blockquote></details>

Why only for tests?
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 74 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHjXpW5t5piHtlcF0u:-OHHjXpW5t5piHtlcF0v:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L74)):*
> ```XML
>         ContractClassManagerConfig { cairo_native_run_config: native_config, ..Default::default() };
>     ContractClassManager::start(config)
> }
> ```





_Suggestion:_
```XML
impl ContractClassManager {
    fn create_for testing(native_config: CairoNativeRunConfig) -> Self {
        let config =
            ContractClassManagerConfig { cairo_native_run_config: native_config, ..Default::default() };
        ContractClassManager::start(config)
    }
}
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 85 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHjm7gDQ9aAEFbBLaS:-OHHjm7gDQ9aAEFbBLaT:bhjqkge) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L85)):*
> ```XML
>     let manager = create_test_manager(native_config);
> 
>     assert_eq!(manager.cairo_native_run_config, native_config);
> ```

Why is it needed?



_Code quote:_
```XML
assert_eq!(manager.cairo_native_run_config, native_config);
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 115 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHkwLh9ea45M5CGRe1:-OHHkwLh9ea45M5CGRe2:b2p06ge) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L115)):*
> ```XML
> #[should_panic(expected = "Compilation request channel is closed.")]
> fn test_send_compilation_request_channel_disconnected() {
>     let native_config = CairoNativeRunConfig { run_cairo_native: true, ..Default::default() };
> ```

I know that it's the default value but it's maybe better to be explicit.



_Suggestion:_
```XML
    // We use the channel to send native compilation requests. 
    let native_config = CairoNativeRunConfig { run_cairo_native: true, wait_on_native_compilation: false, channel_size:1 };
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 155 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHnjYz4TUYFznARfUP:-OHHnjYz4TUYFznARfUQ:b2nhkyh) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L155)):*
> ```XML
>     let native_config = CairoNativeRunConfig {
>         run_cairo_native: true,
>         wait_on_native_compilation: true,
> ```

No?



_Suggestion:_
```XML
wait_on_native_compilation: false,
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 162 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHnzKE8Yrfp4ayPBjw:-OHHnzKE8Yrfp4ayPBjx:bc7er5e) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L162)):*
> ```XML
>     let second_request = create_test_request();
> 
>     // Fill the channel (it can only hold 1 message)
> ```

Where can I see it? Where do you set the channel size to 1? Isn't it 0?



_Code quote:_
```XML
// Fill the channel (it can only hold 1 message)
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 164 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHnmXc2FW3InuDdsYx:-OHHnmXc2FW3InuDdsYy:b-h5faq8) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L164)):*
> ```XML
>     // Fill the channel (it can only hold 1 message)
>     manager.send_compilation_request(request);
>     // Should log an error without panicking
> ```

Can we see the error?



_Code quote:_
```XML
// Should log an error without panicking
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 180 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHpkj720FVPtMMNI3k:-OHHpkj720FVPtMMNI3l:b-uqveti) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L180)):*
> ```XML
>     manager.send_compilation_request(request);
> }
> ```

Consider testing a happy flow of `send_compilation_request` with no wait on native and make sure the channel size was increased by 1.
___
*[`crates/blockifier/src/test_utils/contracts.rs` line 211 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHidT6AJtCw_KbRvl-:-OHHidT6AJtCw_KbRvl0:b-sbbe02) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/test_utils/contracts.rs#L211)):*
> ```XML
>         let raw_sierra = self.get_raw_sierra();
>         serde_json::from_str(&raw_sierra).unwrap()
>     }
> ```

Why was it added? Where do you use it?



_Code quote:_
```XML
    pub fn get_contract_class(&self) -> CairoLangContractClass {
        let raw_sierra = self.get_raw_sierra();
        serde_json::from_str(&raw_sierra).unwrap()
    }
```


<!-- Sent from Reviewable.io -->

noaov1 reviewed (changes requested):
> Reviewed 1 of 13 files at r29.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 11 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @meship-starkware)

___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 197 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHs2jxDlCl6K-QAhFd:-OHHs2jxDlCl6K-QAhFe:b45h) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L197)):*
> ```XML
>     let compiler_config = SierraCompilationConfig::default();
>     let compiler = Arc::new(CommandLineCompiler::new(compiler_config));
>     process_compilation_request(manager.contract_caches.clone(), compiler.clone(), request.clone());
> ```





_Suggestion:_
```XML
    let manager = create_test_manager(native_config);
    let (request, _native) = request_w_native;
    process_compilation_request(manager.contract_caches.clone(), manager.compiler.clone(), request.clone());
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 207 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHsryh5jhvaqLmpHlX:-OHHsryh5jhvaqLmpHlY:bd2d3t5) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L207)):*
> ```XML
> 
> #[rstest]
> fn test_run_compilation_worker() {
> ```

What exactly do we test here? How is it different from testing`process_compilation_request`?



_Code quote:_
```XML
fn test_run_compilation_worker() {
```


<!-- Sent from Reviewable.io -->

noaov1 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 13 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @meship-starkware)

___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 50 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHxSLWD8TEMLdmw0Kk:-OHHxSLXEZlBqga-7ewC:b32w12) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L50)):*
> ```XML
> 
> fn create_test_request() -> CompilationRequest {
>     // Question (AvivG): are we interested in testing other contracts?
> ```

No



_Code quote:_
```XML
// Question (AvivG): are we interested in testing other contracts?
```
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 53 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHyKDB4KA4x97SWApi:-OHHyKDB4KA4x97SWApj:b-ud1bst) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L53)):*
> ```XML
>     let test_contract = FeatureContract::TestContract(CairoVersion::Cairo1(RunnableCairo1::Native));
>     create_test_request_from_contract(test_contract)
> }
> ```

I think that it makes more sense to create the request using the casm. 
Then you don't need the `get_casm` method. 
As for the native:
You don't actually need it. You don't need to check here if the casm is the one that matches the native, you can simply test that you get some native contract. 
WDYT?



_Code quote:_
```XML
fn create_test_request() -> CompilationRequest {
    // Question (AvivG): are we interested in testing other contracts?
    let test_contract = FeatureContract::TestContract(CairoVersion::Cairo1(RunnableCairo1::Native));
    create_test_request_from_contract(test_contract)
}
```


<!-- Sent from Reviewable.io -->

noaov1 reviewed (commented):
> Reviewed 1 of 14 files at r27, 9 of 13 files at r29, 3 of 4 files at r30.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 14 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @meship-starkware)

___
*[`crates/blockifier/src/state/contract_class_manager.rs` line 217 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHzPfY3kFalpe10ohV:-OHHzPfZ8U9GzGSel3RF:b-ahuzne) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager.rs#L217)):*
> ```XML
>             "Contract class with hash {} is already compiled to native. Skipping compilation.",
>             class_hash
>         );
> ```

Consider removing it.



_Code quote:_
```XML
        log::debug!(
            "Contract class with hash {} is already compiled to native. Skipping compilation.",
            class_hash
        );
```


<!-- Sent from Reviewable.io -->

meship-starkware reviewed (approved):
https://github.com/starkware-libs/sequencer/pull/3335 - PR that adds log testing to this PR 

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-:-OHIQdIg4Jxeq3mnzgvp:b-ymxdv8)* status: all files reviewed, 14 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @noaov1)

___
*[`crates/blockifier/src/state/contract_class_manager.rs` line 217 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHzPfY3kFalpe10ohV:-OHINfDKCdfgkkMgfsxH:b205mn7) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager.rs#L217)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Consider removing it.
</blockquote></details>

Maybe adding it under the test config is a good indication for one of the logger tests. I can do this on my PR, which tests the logs.  
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 155 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHnjYz4TUYFznARfUP:-OHIPvFjAyDjV1YRlEl1:b10jkx7) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L155)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

No?
</blockquote></details>

Also fixed in the next PR, actually aviv you can delete this test as this test only has meaning in the log test PR  
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 162 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHnzKE8Yrfp4ayPBjw:-OHIOQeK7Lc-9wlHZX9F:b-t2d9gf) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L162)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Where can I see it? Where do you set the channel size to 1? Isn't it 0?
</blockquote></details>

It will be fixed in the next PR. My splitting was not perfect 
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 164 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHnmXc2FW3InuDdsYx:-OHIOYq_9HxCB7N-ksnP:brn9qof) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L164)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Can we see the error?
</blockquote></details>

yes next PR 
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 180 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHpkj720FVPtMMNI3k:-OHIP8KK7eKxXjG5pMWU:beh0rax) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L180)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Consider testing a happy flow of `send_compilation_request` with no wait on native and make sure the channel size was increased by 1.
</blockquote></details>

I added a log, but that might be better 


<!-- Sent from Reviewable.io -->

noaov1 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 14 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @meship-starkware)

___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 53 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHyKDB4KA4x97SWApi:-OHIa_Nd8PFzVtRgS1uL:bn7hcgm) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L53)):*
<details><summary><i>Previously, avi-starkware (Avi Cohen) wrote…</i></summary><blockquote>

How can you create a request with the casm and not the sierra?
The request is a request to compile sierra to native
</blockquote></details>

You can get the sierra when using casm as the `RunnableCairo1`. I meant use: 
`let test_contract = FeatureContract::TestContract(CairoVersion::Cairo1(RunnableCairo1::Casm));`


<!-- Sent from Reviewable.io -->

noaov1 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 14 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @meship-starkware)

___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 53 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHyKDB4KA4x97SWApi:-OHJBObxAzQXZbiz7LBS:b-25u2nd) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L53)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

You can get the sierra when using casm as the `RunnableCairo1`. I meant use: 
`let test_contract = FeatureContract::TestContract(CairoVersion::Cairo1(RunnableCairo1::Casm));`
</blockquote></details>

I want to avoid using the new method `get_casm`.


<!-- Sent from Reviewable.io -->

meship-starkware reviewed (commented):
> Reviewed 12 of 13 files at r31, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 15 of 16 files reviewed, 12 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @noaov1)

___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 51 at r31](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHc87Hn0uM0qOyGmTc4:-OHcFYY-4pDMPqRmNhDd:bn31qc4) ([raw file](https://github.com/starkware-libs/sequencer/blob/641f66e8b6a0332512f5b8568c229f81d8c4a24c/crates/blockifier/src/state/contract_class_manager_test.rs#L51)):*
<details><summary><i>Previously, avivg-starkware wrote…</i></summary><blockquote>

2 suggestions for avoiding the use of fn get_casm - preference? @noaov1
</blockquote></details>

I would prefer that the test contract would be casm and that the function that would return the runnable class will create a new native contract class 
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 37 at r31](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHcE_EOD_sgdQroBcWa:-OHcE_EP8NmrgjVpq6u9:bnfo9h9) ([raw file](https://github.com/starkware-libs/sequencer/blob/641f66e8b6a0332512f5b8568c229f81d8c4a24c/crates/blockifier/src/state/contract_class_manager_test.rs#L37)):*
> ```XML
>         ContractClassManager::start(config)
>     }
> }
> ```

Move this to the contract class manager file see https://github.com/starkware-libs/sequencer/blob/c29f208b874b1b7ba74e94780c3d68410f1bc665/crates/blockifier/src/blockifier/config.rs#L101



_Code quote:_
```XML
impl ContractClassManager {
    fn create_for_testing(native_config: CairoNativeRunConfig) -> Self {
        let config = ContractClassManagerConfig {
            cairo_native_run_config: native_config,
            ..Default::default()
        };
        ContractClassManager::start(config)
    }
}
```


<!-- Sent from Reviewable.io -->

meship-starkware reviewed (approved):
> Reviewed 1 of 13 files at r31.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 12 unresolved discussions (waiting on @avivg-starkware and @noaov1)



<!-- Sent from Reviewable.io -->

meship-starkware reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 12 unresolved discussions (waiting on @avivg-starkware and @noaov1)



<!-- Sent from Reviewable.io -->

noaov1 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: all files reviewed, 5 unresolved discussions (waiting on @avivg-starkware)

___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 114 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSGWvR9fGFhM4TzBpo:-OHgE5CCDCt_s9_15xke:b3kqx) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/contract_class_manager_test.rs#L114)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Summerizing the above:
`process_compilation_request`- tested in Meshi's PR (no need to distinguish between the different values of `wait_on_native`.
`send_compilation_request`-tested in Meshi's PR (*need* to distinguish between the different values of `wait_on_native`.
Please correct me if I'm wrong.
</blockquote></details>

.


<!-- Sent from Reviewable.io -->

avivg-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 2 of 17 files reviewed, 4 unresolved discussions (waiting on @avi-starkware, @meship-starkware, and @noaov1)

___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 114 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OFSGWvR9fGFhM4TzBpo:-OHgf_gfDAZj19yX56Jp:bmeyrxb) ([raw file](https://github.com/starkware-libs/sequencer/blob/9547c32e33c4c2c3ba7f9afcca3b457892ce2edb/crates/blockifier/src/state/contract_class_manager_test.rs#L114)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

.
</blockquote></details>

:+1:  
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 207 at r30](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHHsryh5jhvaqLmpHlX:-OHggjxp1BWaNcdyMztJ:b-eezgdl) ([raw file](https://github.com/starkware-libs/sequencer/blob/ce8f9a1648854acabc13341e192eeb3d194667d0/crates/blockifier/src/state/contract_class_manager_test.rs#L207)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

How exactly do you test it? Simply by checking that the test was finished?
</blockquote></details>

test removed
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 37 at r31](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHcE_EOD_sgdQroBcWa:-OHgD-RyBVz-iV5S7Tbv:b-fu5a33) ([raw file](https://github.com/starkware-libs/sequencer/blob/641f66e8b6a0332512f5b8568c229f81d8c4a24c/crates/blockifier/src/state/contract_class_manager_test.rs#L37)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

Move this to the contract class manager file see https://github.com/starkware-libs/sequencer/blob/c29f208b874b1b7ba74e94780c3d68410f1bc665/crates/blockifier/src/blockifier/config.rs#L101
</blockquote></details>

moved to  "struct_impls.rs"  
___
*[`crates/blockifier/src/state/contract_class_manager_test.rs` line 51 at r31](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-OHc87Hn0uM0qOyGmTc4:-OHggpjlASs9FubIqt5-:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/641f66e8b6a0332512f5b8568c229f81d8c4a24c/crates/blockifier/src/state/contract_class_manager_test.rs#L51)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

I would prefer that the test contract would be casm and that the function that would return the runnable class will create a new native contract class
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

noaov1 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 7 files at r32, 12 of 12 files at r33, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768#-:-OHgmCb-9US3eyOKRQkh:bnfp4nl)* status: 13 of 17 files reviewed, 1 unresolved discussion (waiting on @meship-starkware)



<!-- Sent from Reviewable.io -->

meship-starkware reviewed (commented):
> Reviewed 1 of 7 files at r32, 12 of 12 files at r33, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: 13 of 17 files reviewed, all discussions resolved (waiting on @noaov1)



<!-- Sent from Reviewable.io -->

meship-starkware reviewed (approved):
> Reviewed 4 of 7 files at r32.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2768)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @avivg-starkware)



<!-- Sent from Reviewable.io -->
